File: src-tauri/src/commands/activity.rs
use crate::AppState;
use crate::services::activity::{
  ActivityLogItem, DailyProgress, UserGoal, get_activity_log, get_streak_count, get_today_progress,
  get_user_goals, record_activity, update_user_goals,
};
use tauri::State;

#[tauri::command]
pub async fn record_daily_activity(
  state: State<'_, AppState>,
  char_diff: i64,
) -> Result<(), String> {
  let context = state.get_context()?;
  let conn = context.db.conn.lock().map_err(|e| e.to_string())?;
  record_activity(&conn, char_diff).map_err(|e| e.to_string())?;
  Ok(())
}

#[tauri::command]
pub async fn get_streak(state: State<'_, AppState>) -> Result<i64, String> {
  let context = state.get_context()?;
  let conn = context.db.conn.lock().map_err(|e| e.to_string())?;
  let streak = get_streak_count(&conn).map_err(|e| e.to_string())?;
  Ok(streak)
}

#[tauri::command]
pub async fn get_activity_heatmap(
  state: State<'_, AppState>,
) -> Result<Vec<ActivityLogItem>, String> {
  let context = state.get_context()?;
  let conn = context.db.conn.lock().map_err(|e| e.to_string())?;
  let logs = get_activity_log(&conn).map_err(|e| e.to_string())?;
  Ok(logs)
}

#[tauri::command]
pub async fn get_goals(state: State<'_, AppState>) -> Result<UserGoal, String> {
  let context = state.get_context()?;
  let conn = context.db.conn.lock().map_err(|e| e.to_string())?;
  let goals = get_user_goals(&conn).map_err(|e| e.to_string())?;
  Ok(goals)
}

#[tauri::command]
pub async fn set_goals(state: State<'_, AppState>, goals: UserGoal) -> Result<(), String> {
  let context = state.get_context()?;
  let conn = context.db.conn.lock().map_err(|e| e.to_string())?;
  update_user_goals(&conn, goals).map_err(|e| e.to_string())?;
  Ok(())
}

#[tauri::command]
pub async fn get_daily_progress(state: State<'_, AppState>) -> Result<DailyProgress, String> {
  let context = state.get_context()?;
  let conn = context.db.conn.lock().map_err(|e| e.to_string())?;
  let progress = get_today_progress(&conn).map_err(|e| e.to_string())?;
  Ok(progress)
}

File: src/stores/activity.ts
import { create } from 'zustand';
import { type ActivityLogItem, type DailyProgress, type UserGoal } from '@/types/activity';
import { invoke } from '@tauri-apps/api/core';

type ActivityState = {
  streak: number;
  activityLog: ActivityLogItem[];
  goals: UserGoal | null;
  dailyProgress: DailyProgress | null;
  isLoading: boolean;
  error: string | null;

  // Actions
  fetchStreak: () => Promise<void>;
  fetchActivityLog: () => Promise<void>;
  fetchGoals: () => Promise<void>;
  updateGoals: (goals: UserGoal) => Promise<void>;
  fetchDailyProgress: () => Promise<void>;
  recordActivity: (charDiff: number) => Promise<void>;
};

export const useActivityStore = create<ActivityState>((set, get) => ({
  streak: 0,
  activityLog: [],
  goals: null,
  dailyProgress: null,
  isLoading: false,
  error: null,

  fetchStreak: async () => {
    try {
      set({ isLoading: true, error: null });
      const streak = await invoke<number>('get_streak');
      set({ streak, isLoading: false });
    } catch (error) {
      set({ error: String(error), isLoading: false });
    }
  },

  fetchActivityLog: async () => {
    try {
      set({ isLoading: true, error: null });
      const activityLog = await invoke<ActivityLogItem[]>('get_activity_heatmap');
      set({ activityLog, isLoading: false });
    } catch (error) {
      set({ error: String(error), isLoading: false });
    }
  },

  fetchGoals: async () => {
    try {
      set({ isLoading: true, error: null });
      const goals = await invoke<UserGoal>('get_goals');
      set({ goals, isLoading: false });
    } catch (error) {
      set({ error: String(error), isLoading: false });
    }
  },

  updateGoals: async (goals: UserGoal) => {
    try {
      set({ isLoading: true, error: null });
      await invoke('set_goals', { goals });
      set({ goals, isLoading: false });
    } catch (error) {
      set({ error: String(error), isLoading: false });
    }
  },

  fetchDailyProgress: async () => {
    try {
      set({ isLoading: true, error: null });
      const dailyProgress = await invoke<DailyProgress>('get_daily_progress');
      set({ dailyProgress, isLoading: false });
    } catch (error) {
      set({ error: String(error), isLoading: false });
    }
  },

  recordActivity: async (charDiff: number) => {
    try {
      await invoke('record_daily_activity', { charDiff });
      // Refresh streak and daily progress after recording
      await get().fetchStreak();
      await get().fetchDailyProgress();
    } catch (error) {
      set({ error: String(error) });
    }
  }
}));
